### PR TITLE
Combines standalone chars into text groups.

### LINKED ISSUE #1045: 100% CPU due to text layer generating one div per letter
Here is the PDF
http://www.music.mcgill.ca/~cmckay/papers/musictech/ISMIR_2006_Genre.pdf

Browser is Chrome 16
OS is Ubuntu 10.10

### PR DESCRIPTION
**Pull request by yurydelendik**
Fixes #1045

TODOs:
- [x] land #6588 and rebase
- [x] add fake spaces when advance moved a lot (e.g. /test/pdfs/fips197.pdf)
- [x] <del>add tail padding (?) see #5972 (out of scope)</del>
- [x] test

### CODE REVIEW

File: src/core/evaluator.js [hunk 1]
```diff
@@ -1119,35 +1153,60 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+              flushTextContentItem();
               textState.textHScale = args[0] / 100;
               break;
             case OPS.setLeading:
+              flushTextContentItem();
               textState.leading = args[0];
               break;
             case OPS.moveText:
+              // Optimization to treat same line movement as advance
+              var isSameTextLine = textState.font.vertical ?
```
Comment (Snuffleupagus) [inline src/core/evaluator.js]: This line seems to cause `uncaught exception: TypeError: textState.font is null` when opening the PDF file from https://bugzilla.mozilla.org/show_bug.cgi?id=960315.

**Comment by timvandermeij** [comment]
The text layer is appearing almost instantly now for #1045 and memory usage also appears to be lower, although I could not get a reliable measurement. Really nice work!

**Comment by existentialism** [comment]
:+1: this is fantastic work! I've been testing with some of our more problematic PDFs and seeing huge improvements.

**Comment by yurydelendik** [comment]
TODO:
- [ ] firefox-tracemonkey-text-page3 - suppress breakTextRun logic for monospace fonts
- [ ] firefox-issue3925-page1 - some runs can collapse into zero width
- [ ] firefox-taro-text-page1 - vertical text issues

**Comment by yurydelendik** [comment]
(How report amount of div in a text layer: `document.querySelectorAll('.page')[/* pageIndex */ 0].querySelectorAll('.textLayer > div').length`;

File: src/core/evaluator.js [hunk 2]
```diff
@@ -1119,35 +1174,62 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+              advance = args[0] - args[1];
+              if (isSameTextLine && textContentItem.initialized &&
+                  advance > 0 &&
+                  advance <= textContentItem.fakeMultiSpaceMax) {
+                textState.translateTextLineMatrix(args[0], args[1]);
+                textContentItem.width +=
+                  (args[0] - textContentItem.lastAdvanceWidth);
+                textContentItem.height +=
+                  (args[1] - textContentItem.lastAdvanceHeight);
+                var diff = (args[0] - textContentItem.lastAdvanceWidth) -
```
Comment (brendandahl) [inline src/core/evaluator.js]: Why are width and height used here?

Comment (brendandahl) [inline src/core/evaluator.js]: Oh, I guess if we're on the same line one will cancel out to zero?

Comment (yurydelendik) [inline src/core/evaluator.js]: that's correct.

**Comment by speedplane** [comment]
I believe this PR introduced some issues with horizontal text scaling.


image

**Comment by yurydelendik** [comment]
> I believe this PR introduced some issues with horizontal text scaling.

Please provide details for OS and browsers you are using. lgtm on Firefox for Mac OSX

**Comment by speedplane** [comment]
Just upgraded firefox from 41 to 42 and this issue went away. Very strange.

**Comment by Snuffleupagus** [comment]
For future reference, this also fixed https://bugzilla.mozilla.org/show_bug.cgi?id=1136674.

**Pull request merged**